src/cdk-experimental/testing/component-harness.ts

```diff
@@ -8,183 +8,267 @@
 
 import {TestElement} from './test-element';
 
-/** Options that can be specified when querying for an Element. */
-export interface QueryOptions {
-  /**
-   * Whether the found element can be null. If allowNull is set, the searching function will always
-   * try to fetch the element at once. When the element cannot be found, the searching function
-   * should return null if allowNull is set to true, throw an error if allowNull is set to false.
-   * If allowNull is not set, the framework will choose the behaviors that make more sense for each
-   * test type (e.g. for unit test, the framework will make sure the element is not null; otherwise
-   * throw an error); however, the internal behavior is not guaranteed and user should not rely on
-   * it. Note that in most cases, you don't need to care about whether an element is present when
-   * loading the element and don't need to set this parameter unless you do want to check whether
-   * the element is present when calling the searching function. e.g. you want to make sure some
-   * element is not there when loading the element in order to check whether a "ngif" works well.
-   */
-  allowNull?: boolean;
-  /**
-   * If global is set to true, the selector will match any element on the page and is not limited to
-   * the root of the harness. If global is unset or set to false, the selector will only find
-   * elements under the current root.
-   */
-  global?: boolean;
-}
+/** An async function that returns a promise when called. */
+export type AsyncFn<T> = () => Promise<T>;
 
-/** Interface that is used to find elements in the DOM and create harnesses for them. */
-export interface HarnessLocator {
+/**
+ * Interface used to load ComponentHarness objects. This interface is used by test authors to
+ * instantiate `ComponentHarness`es.
+ */
+export interface HarnessLoader {
   /**
-   * Get the host element of locator.
+   * Searches for an element with the given selector under the current instances's root element,
+   * and returns a `HarnessLoader` rooted at the matching element. If multiple elements match the
+   * selector, the first is used. If no elements match, an error is thrown.
+   * @param selector The selector for the root element of the new `HarnessLoader`
+   * @return A `HarnessLoader` rooted at the element matching the given selector.
+   * @throws If a matching element can't be found.
    */
-  host(): TestElement;
+  findRequired(selector: string): Promise<HarnessLoader>;
```

Comment:
We can defer final decision to the team meeting for consensus, but my own preference would be for `find` and `findOptional`, where the former is "required" by default. (Similar for other uses of "required")

Reply:
I left the `required` on there for now to leave the door open for adding a plain `find` method. This would be the equivalent of `required` for unit tests, but for e2e tests it wouldn't bother to check `element.isPresent` and would just assume the element is there. This is what happens in the current API when you don't specify anything for the `allowNull` option. A comment mentions doing this for performance reasons, but I wanted to see if anyone had more context on whether we should have this option or not.